File: app/src/main/java/com/inipage/homelylauncher/dock/DockController.java
package com.inipage.homelylauncher.dock;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.Nullable;

import com.inipage.homelylauncher.R;
import com.inipage.homelylauncher.dock.items.AlarmMappedDockItem;
import com.inipage.homelylauncher.dock.items.CalendarMappedDockItem;
import com.inipage.homelylauncher.dock.items.ContextualAppFetcher;
import com.inipage.homelylauncher.dock.items.PhoneMappedDockItem;
import com.inipage.homelylauncher.dock.items.PowerMappedDockItem;
import com.inipage.homelylauncher.dock.items.WeatherDockItem;
import com.inipage.homelylauncher.model.DockItem;
import com.inipage.homelylauncher.persistence.DatabaseEditor;
import com.inipage.homelylauncher.utils.ViewUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

/**
 * Manages dock items at the bottom of the home screen.
 */
public class DockController {

    private static final int MAX_ITEM_COUNT = 20;

    private final LinearLayout mContainer;
    private final ContextualAppFetcher mAppFetcher;

    private List<DockControllerItem> mActiveDockItems;
    private Map<Integer, DockItem> mAppBackedItemsCache;

    public DockController(LinearLayout container) {
        mContainer = container;
        mAppFetcher = new ContextualAppFetcher();
        mAppBackedItemsCache = new HashMap<>();
    }

    public synchronized void loadDock() {
        destroyDockImpl();

        final Context context = mContainer.getContext();
        final boolean isSquarish = ViewUtils.isSquarishDevice(context);

        // Setup the dock controller supporting fields
        mAppFetcher.reloadPrefs();
        mAppBackedItemsCache = DatabaseEditor.get().getDockPreferences()
            .parallelStream()
            .filter(dockItem -> dockItem.getWhenToShow() !=
                DockItem.DOCK_SHOW_NEVER)
            .collect(Collectors.toConcurrentMap(
                DockItem::getWhenToShow,
                Function.identity()));

        // Add all the dock items
        mActiveDockItems = new ArrayList<>();
        mActiveDockItems.add(new AlarmMappedDockItem());
        mActiveDockItems.add(new CalendarMappedDockItem());
        mActiveDockItems.add(new WeatherDockItem());
        mActiveDockItems.add(new PhoneMappedDockItem(mAppBackedItemsCache));
        mActiveDockItems.add(new PowerMappedDockItem(mAppBackedItemsCache));
        mActiveDockItems.addAll(mAppFetcher.getRecentApps(context));
        mActiveDockItems =
            mActiveDockItems
                .parallelStream()
                .sorted((left, right) -> {
                    if (left.getBasePriority() != right.getBasePriority()) {
                        return (int) (right.getBasePriority() - left.getBasePriority());
                    }
                    return (int) (right.getSubPriority() - left.getSubPriority());
                }).collect(Collectors.toList());

        // Bind the dock items to views
        mContainer.addView(ViewUtils.createFillerWidthView(
            context,
            (int) context.getResources().getDimension(
                isSquarish ?
                    R.dimen.home_activity_squarish :
                    R.dimen.home_activity_margin)));
        for (int i = 0; i < Math.min(mActiveDockItems.size(), MAX_ITEM_COUNT); i++) {
            final LinearLayout.LayoutParams layoutParams =
                new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            final View view =
                LayoutInflater.from(context).inflate(
                    R.layout.dock_collapsed_item, mContainer, false);
            view.setVisibility(GONE);
            mContainer.addView(view, layoutParams);

            final DockControllerItem item = mActiveDockItems.get(i);
            item.attach(new DockControllerItem.Host() {
                @Override
                public Context getContext() {
                    return mContainer.getContext();
                }

                @Override
                public void showHostedItem() {
                    bindViewToLoadedItem(isSquarish, view, item);
                    view.setVisibility(VISIBLE);
                }

                @Override
                public void hideHostedItem() {
                    view.setVisibility(GONE);
                }

                @Override
                public void tintLoaded(int color) {
                    final View dockItemContainer = view.findViewById(R.id.dock_item_root_container);
                    dockItemContainer.getBackground().setColorFilter(
                        new PorterDuffColorFilter(
                            color,
                            PorterDuff.Mode.SRC_IN));
                }
            });

        }
        mContainer.addView(ViewUtils.createFillerWidthView(
            context, (int) context.getResources().getDimension(R.dimen.home_activity_margin)));
    }

    public void destroyDock() {
        destroyDockImpl();
    }

    private void destroyDockImpl() {
        mContainer.removeAllViews();
        mAppBackedItemsCache.clear();
        if (mActiveDockItems != null) {
            for (DockControllerItem item : mActiveDockItems) {
                item.detach();
            }
            mActiveDockItems.clear();
            mActiveDockItems = null;
        }
    }

    private static void bindViewToLoadedItem(
            boolean isSquarish,
            View view,
            DockControllerItem item) {
        // Setup background color
        final View dockItemContainer = view.findViewById(R.id.dock_item_root_container);
        dockItemContainer.getBackground().setColorFilter(
            new PorterDuffColorFilter(
                item.getTint(),
                PorterDuff.Mode.SRC_IN));

        // Map icon
        ImageView itemIcon = ((ImageView) view.findViewById(R.id.contextual_dock_item_icon));
        int icon = item.getIcon();
        @Nullable Bitmap bitmap = item.getBitmap();
        @Nullable Drawable drawable = item.getDrawable();
        if (icon != 0) {
            itemIcon.setImageResource(icon);
        } else if (bitmap != null) {
            itemIcon.setImageBitmap(bitmap);
        } else if (drawable != null) {
            itemIcon.setImageDrawable(drawable);
        }

        // Map text
        @Nullable final String label = item.getLabel();
        if (label != null) {
            ((TextView) view.findViewById(R.id.contextual_dock_item_label)).setText(label);
        } else {
            ((View) view.findViewById(R.id.contextual_dock_item_label).getParent())
                .setVisibility(GONE);
        }
        @Nullable final String secondaryLabel = item.getSecondaryLabel();
        if (secondaryLabel != null && !isSquarish) {
            ((TextView) view.findViewById(R.id.contextual_dock_item_secondary_label)).setText(
                secondaryLabel);
        } else {
            view.findViewById(R.id.contextual_dock_item_secondary_label).setVisibility(GONE);
        }

        // Map actions
        view.setOnClickListener(v -> item.getAction(v).run());
        view.setOnLongClickListener(v -> {
            @Nullable final Runnable action = item.getSecondaryAction(v);
            if (action != null) {
                action.run();
            }
            return action != null;
        });
    }
}

File: app/src/main/java/com/inipage/homelylauncher/dock/DockController.kt
package com.inipage.homelylauncher.dock

import android.content.Context
import android.graphics.Rect
import android.os.Handler
import androidx.recyclerview.widget.RecyclerView
import com.inipage.homelylauncher.dock.items.ContextualAppFetcher
import com.inipage.homelylauncher.dock.DockControllerItem
import com.inipage.homelylauncher.model.DockItem
import android.os.Looper
import android.util.Log
import android.view.View
import com.inipage.homelylauncher.persistence.DatabaseEditor
import com.inipage.homelylauncher.dock.items.AlarmMappedDockItem
import com.inipage.homelylauncher.dock.items.CalendarMappedDockItem
import com.inipage.homelylauncher.dock.items.WeatherDockItem
import com.inipage.homelylauncher.dock.items.PhoneMappedDockItem
import com.inipage.homelylauncher.dock.items.PowerMappedDockItem
import androidx.recyclerview.widget.LinearLayoutManager
import com.inipage.homelylauncher.R
import com.inipage.homelylauncher.dock.DockAdapter
import com.inipage.homelylauncher.persistence.PrefsHelper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.ArrayList
import java.util.HashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.function.Function
import java.util.stream.Collectors

/**
 * Manages dock items at the bottom of the home screen.
 */
class DockController(val container: RecyclerView) {

    private val foregroundHandler = Handler(Looper.getMainLooper())
    private val backgroundExecutor = Executors.newFixedThreadPool(4)
    private val appFetcher = ContextualAppFetcher()

    private var adapter: DockAdapter? = null
    private var activeDockItems: MutableList<DockControllerItem> = ArrayList()
    private var appBackedItemsCache: MutableMap<Int, DockItem> = HashMap()
    private var isMono: Boolean = false
    
    fun loadDock() {
        destroyDockImpl()
        backgroundExecutor.submit {
            loadDockItemsImpl()
        }
    }

    fun destroyDock() {
        destroyDockImpl()
    }

    private fun loadDockItemsImpl() {
        // Setup the dock controller supporting fields
        appFetcher.reloadPrefs()
        isMono = PrefsHelper.usingMonochromeDock()
        appBackedItemsCache = DatabaseEditor.get().dockPreferences
            .parallelStream()
            .filter { dockItem: DockItem -> dockItem.whenToShow != DockItem.DOCK_SHOW_NEVER }
            .collect(
                Collectors.toConcurrentMap(
                    { obj: DockItem -> obj.whenToShow },
                    Function.identity()
                )
            )

        // Add all the dock items
        activeDockItems.add(AlarmMappedDockItem())
        activeDockItems.add(CalendarMappedDockItem())
        activeDockItems.add(WeatherDockItem())
        activeDockItems.add(PhoneMappedDockItem(appBackedItemsCache))
        activeDockItems.add(PowerMappedDockItem(appBackedItemsCache))
        // This takes about ~150-200ms on a fast-ish device; the rest of the init takes about
        // 40-50ms
        activeDockItems.addAll(appFetcher.getRecentApps(container.context, true))
        activeDockItems = activeDockItems
            .stream()
            .sorted { left, right ->
                val basePriority = right.basePriority - left.basePriority
                return@sorted (if (basePriority == 0) {
                    right.subPriority - left.subPriority
                } else basePriority)
            }
            .collect(Collectors.toList())

        foregroundHandler.post { attachDockItemsToView() }
    }

    private fun destroyDockImpl() {
        container.adapter = null
        for (item in activeDockItems) {
            item.detach()
        }
        activeDockItems.clear()
        appBackedItemsCache.clear()
    }

    private fun attachDockItemsToView() {
        container.adapter = DockAdapter(container.context, activeDockItems, isMono)
        activeDockItems.forEachIndexed { index, item ->
            backgroundExecutor.submit {
                item.attach(object : DockControllerItem.Host {

                    override fun getContext() = container.context

                    override fun showHostedItem() {
                        foregroundHandler.post {
                            adapter?.notifyItemChanged(index)
                        }
                    }

                    override fun hideHostedItem() {
                        foregroundHandler.post {
                            adapter?.notifyItemChanged(index) ?: Unit
                        }
                    }

                    override fun tintLoaded(color: Int) {
                        foregroundHandler.post {
                            adapter?.notifyItemChanged(index) ?: Unit
                        }
                    }
                })
            }
        }
        container.alpha = 0f
        container.animate()
            .alpha(1f)
            .withStartAction { container.visibility = View.VISIBLE }
            .setDuration(150L)
            .start()
    }

    init {
        val context = container.context
        val betweenItemSpacePx = context.resources.getDimensionPixelSize(R.dimen.contextual_dock_internal_padding)
        container.layoutManager =
            LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        container.addItemDecoration(object : RecyclerView.ItemDecoration() {
            override fun getItemOffsets(
                outRect: Rect,
                view: View,
                parent: RecyclerView,
                state: RecyclerView.State
            ) {
                val idx = container.getChildAdapterPosition(view)
                if (idx == RecyclerView.NO_POSITION) {
                    return
                }
                val isFirstItem = idx == 0
                val isLastItem = idx == ((adapter?.itemCount ?: 0) - 1)
                val isHidden = !(activeDockItems.getOrNull(idx)?.isLoaded ?: false)
                outRect.set(
                    if (isFirstItem) (betweenItemSpacePx * 1.5).toInt() else 0,
                    0,
                    if (isLastItem || isHidden) 0 else betweenItemSpacePx,
                    0)
            }
        })
    }
}